aspnet/6-pubsub/lib/BookDetailLookup.cs

```diff
@@ -223,15 +225,61 @@ public static void ProcessBook(IBookStore bookStore, long bookId)
             var response = WebRequest.Create(query).GetResponse();
             var reader = new StreamReader(response.GetResponseStream());
             var json = reader.ReadToEnd();
-            dynamic results = JObject.Parse(json);
-            dynamic info = results.items[0].volumeInfo;
-            book.Title = info.title;
-            book.PublishedDate = info.publishedDate;
-            book.Author = string.Join(", ", info.authors);
-            book.Description = info.description;
+            ParseBook(json, ref book);
             // TODO: update image.
             bookStore.Update(book);
         }
         // [END processbook]
+
+        /// <summary>
+        /// Parse a date time.  Return null if it can't be parsed.  A single number will be 
+        /// interpreted as a year, and the date returned will be YEAR-01-01.
+        /// </summary>
+        /// <param name="dateString">A string representation of the date.</param>
+        /// <returns>A datetime or null.</returns>
+        public static DateTime? ParseDate(string dateString)
+        {
+            DateTime result;
+            if (DateTime.TryParse(dateString, out result))
+                return result;
+            int year;
+            if (int.TryParse(dateString, out year))
+                return new DateTime(year, 1, 1);
+            return null;
+        }
+
+        /// <summary>
+        /// Updates book with information parsed from a json response from Google's Books API.
+        /// </summary>
+        /// <param name="json">A response from Google's Books API.</param>
+        /// <param name="book">Fields in book will be overwritten.</param>
+        public static void ParseBook(string json, ref Book book)
+        {
+            // There are many volumeInfos, and many are incomplete.  So, to find a field like
+            // "title", we use LINQ to scan multiple volumeInfos.
+            JObject results = JObject.Parse(json);
+            var infos = results["items"].Select(token => (JObject)token)
+                .Where(obj => obj.Property("volumeInfo") != null)
+                .Select(obj => obj["volumeInfo"]);
+            Func<string, IEnumerable<JToken>> GetInfo = (propertyName) =>
+            {
+                return infos.Select(token => (JObject)token)
+                    .Where(info => info.Property(propertyName) != null)
+                    .Select(info => info[propertyName]);
+            };
+            foreach (var title in GetInfo("title").Take(1))

```

Comment:
Yes.

On Tue, Mar 15, 2016 at 12:03 PM, remi Taylor <notifications@github.com>
wrote:

> In aspnet/6-pubsub/lib/BookDetailLookup.cs
> <https://github.com/GoogleCloudPlatform/getting-started-dotnet/pull/30#discussion_r56221970>
> :
>
> > +        /// <param name="book">Fields in book will be overwritten.</param>
> > +        public static void ParseBook(string json, ref Book book)
> > +        {
> > +            // There are many volumeInfos, and many are incomplete.  So, to find a field like
> > +            // "title", we use LINQ to scan multiple volumeInfos.
> > +            JObject results = JObject.Parse(json);
> > +            var infos = results["items"].Select(token => (JObject)token)
> > +                .Where(obj => obj.Property("volumeInfo") != null)
> > +                .Select(obj => obj["volumeInfo"]);
> > +            Func<string, IEnumerable<JToken>> GetInfo = (propertyName) =>
> > +            {
> > +                return infos.Select(token => (JObject)token)
> > +                    .Where(info => info.Property(propertyName) != null)
> > +                    .Select(info => info[propertyName]);
> > +            };
> > +            foreach (var title in GetInfo("title").Take(1))
>
> Does this mean that GetInfo("title") might return a title from a
> *different* volumeInfo than GetInfo("authors") or GetInfo("publishedDate")
> etc might return?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/getting-started-dotnet/pull/30/files/f0362338f5dcf8caa50ebb090ccb9c449e8fa805#r56221970>
>